packages/@lwc/engine/src/framework/vm.ts

```diff
@@ -686,12 +699,29 @@ export function allocateInSlot(vm: VM, children: VNodes) {
             const oldVNodes = oldSlots[key];
             const vnodes = cmpSlots[key];
             for (let j = 0, a = cmpSlots[key].length; j < a; j += 1) {
-                if (oldVNodes[j] !== vnodes[j]) {
+                if (hasDifferentIdentities(oldVNodes[j], vnodes[j])) {
                     markComponentAsDirty(vm);
                     return;
                 }
             }
         }
+        // If we to this point in the algo, it means that the allocation children
+        // collection identical, no new nodes or removal, and all existing vnodes
+        // have the same identity (key). We can just simply diff those vnodes directly
+        // to avoid marking the vm as dirty, which will cause a full rehydration
+        // just to allocate the exact same things.
+        const { aChildren } = vm;
+        const slottedNewChildren = [];
+        const slottedOldChildren = [];
+        for (let i = 0, len = children.length; i < len; i += 1) {
+            const oldVNode = aChildren[i];
+            if (!isNull(oldVNode) && !isUndefined(oldVNode.elm)) {
+                const newVNode = children[i];
+                ArrayPush.call(slottedNewChildren, oldVNode);
+                ArrayPush.call(slottedOldChildren, newVNode);
```

Comment:
do we use `newVNode` elsewhere?

Reply:
yes, we can inline it, but the meaning of it is important, that's why I hoist it